ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/Deduplicate.java

```diff
@@ -4,128 +4,169 @@
-      ResultWithErrors<PCollection<PubsubMessage>> result = input.apply(new RemoveDuplicates(uri));
-      PCollection<PubsubMessage> errorsStrippedOfDuplicates = result.errors()
-          .apply(Filter.by((PubsubMessage message) -> !Optional
-              .ofNullable(message.getAttribute("exception_class"))
-              .filter(RemoveDuplicates.DuplicateIdException.class.getName()::equals).isPresent()));
-      return ResultWithErrors.of(result.output(), errorsStrippedOfDuplicates);
-    });
+  public static RemoveDuplicates removeDuplicates(ValueProvider<URI> uri) {
+    return new RemoveDuplicates(uri);
   }
 
-  public static Deduplicate markAsSeen(ValueProvider<URI> uri, ValueProvider<Integer> ttlSeconds) {
+  /**
+   * Returns a {@link PTransform} that extracts document_id from each message and marks each id as
+   * seen in Redis.
+   */
+  public static MarkAsSeen markAsSeen(ValueProvider<URI> uri, ValueProvider<Integer> ttlSeconds) {
     return new MarkAsSeen(uri, ttlSeconds);
   }
 
-  /*
-   * Concrete subclasses.
-   */
-
   /**
-   * {@link PTransform} that redirects messages already seen to {@code errorTag}.
+   * Implementation of {@link #removeDuplicates(ValueProvider)}.
    */
-  private static class RemoveDuplicates extends Deduplicate {
+  public static class RemoveDuplicates extends PTransform<PCollection<PubsubMessage>, RemoveDuplicates.Result> {
+
+    private final RedisIdService redisIdService;
 
-    private RemoveDuplicates(ValueProvider<URI> uri) {
-      super(uri);
+    private final TupleTag<PubsubMessage> outputTag = new TupleTag<PubsubMessage>() {
+    };
+    private final TupleTag<PubsubMessage> duplicateTag = new TupleTag<PubsubMessage>() {
+    };
+    private final TupleTag<PubsubMessage> errorTag = new TupleTag<PubsubMessage>() {
+    };
+
+    RemoveDuplicates(ValueProvider<URI> uri) {
+      this.redisIdService = new RedisIdService(uri);
     }
 
-    private static class DuplicateIdException extends Exception {
+    @AutoValue
+    public abstract static class Result implements PInput, POutput {
+      public abstract PCollectionTuple tuple();
+      public abstract TupleTag<PubsubMessage> outputTag();
+      public abstract TupleTag<PubsubMessage> duplicateTag();
+      public abstract TupleTag<PubsubMessage> errorTag();
+
+      public static Result of(PCollectionTuple tuple,
+          TupleTag<PubsubMessage> outputTag, TupleTag<PubsubMessage> duplicateTag,
+          TupleTag<PubsubMessage> errorTag) {
+        return new AutoValue_Deduplicate_RemoveDuplicates_Result(tuple, outputTag, duplicateTag, errorTag);
+      }
+
+      /**
+       * Ignore the collection of duplicates and return a {@link Result} of just the
+       * non-duplicate output and the errors.
+       */
+      public WithErrors.Result<PCollection<PubsubMessage>> ignoreDuplicates() {
+        return WithErrors.Result.of(tuple().get(outputTag()), outputTag(), tuple().get(errorTag()), errorTag());
```

Comment:
I don't see where `WithErrors.Result.of` is defined to tag pairs of pcollections and tags. could you explain to me what's happening here?